plugins/ltac/extratactics.mlg

```diff
@@ -1119,3 +1119,16 @@ let tclOPTIMIZE_HEAP =
 TACTIC EXTEND optimize_heap
 | [ "optimize_heap" ] -> { tclOPTIMIZE_HEAP }
 END
+
+(** tactic analogous to "Strategy" *)
+
+TACTIC EXTEND strategy
+| [ "strategy" strategy_level_or_var(v) "[" ne_smart_global_list(q) "]"  ] -> {
+  Proofview.Goal.enter begin
+    fun _ ->
+      let local = true in
+      Vernacentries.set_strategy ~local [(v, q)];
```

Comment:
This is extremely fragile and likely a piece of code that is going to break when we move towards a more functional implementation. This function is actually no more than a wrapper around its `Global` counterpart, as the VM-acting function in `Vernacentries` is a huge bluff, it's just a no-op. Thus this should probably be defined by tweaking the tactic-local environment with `tclENV` instead, and pray that no one is using the imperative global environment in the tactics you're interested in.

Reply:
Does `tclENV` play nicely with `abstract`?  Note the following in the test-suite:
```coq
(* test that the strategy tactic does persist through abstract *)
Opaque id.
Goal id 0 = 0.
  strategy expand [id].
  assert (id (fact 100) = fact 100) by abstract reflexivity.
  reflexivity.
Defined.
```

> This function is actually no more than a wrapper around its `Global` counterpart, as the VM-acting function in `Vernacentries` is a huge bluff, it's just a no-op.

What is the "VM-acting function" you refer to?  Is it the `Csymtable.set_opaque_const` / `Csymtable.set_transparent_const` call?  Or the call to `Lib.add_anonymous_leaf`?

Are you suggesting we instead do something like
```ocaml
  let glob_key r =
    match r with
      | GlobRef.ConstRef sp -> ConstKey sp
      | GlobRef.VarRef id -> VarKey id
      | _ -> user_err Pp.(str
          "cannot set an inductive type or a constructor as transparent") in
  tclENV >>= fun env ->
  let env = List.fold_left (fun env q -> Environ.set_oracle env (Conv_oracle.set_strategy (Environ.oracle env) (glob_key q) v)) env ql in
  tclSETENV env
```
?  (And if so, which `.ml` file should this live in?)

Reply:
Does `tclENV` play nicely with `abstract`?  Note the following in the test-suite:
```coq
(* test that the strategy tactic does persist through abstract *)
Opaque id.
Goal id 0 = 0.
  strategy expand [id].
  assert (id (fact 100) = fact 100) by abstract reflexivity.
  reflexivity.
Defined.
```

> This function is actually no more than a wrapper around its `Global` counterpart, as the VM-acting function in `Vernacentries` is a huge bluff, it's just a no-op.

What is the "VM-acting function" you refer to?  Is it the `Csymtable.set_opaque_const` / `Csymtable.set_transparent_const` call?  Or the call to `Lib.add_anonymous_leaf`?

Are you suggesting we instead do something like
```ocaml
  let glob_key r =
    match r with
      | GlobRef.ConstRef sp -> ConstKey sp
      | GlobRef.VarRef id -> VarKey id
      | _ -> user_err Pp.(str
          "cannot set an inductive type or a constructor as transparent") in
  tclENV >>= fun env ->
  let env = List.fold_left (fun env q -> Environ.set_oracle env (Conv_oracle.set_strategy (Environ.oracle env) (glob_key q) v)) env ql in
  tclSETENV env
```
?  (And if so, which `.ml` file should this live in?)

Reply:
> Does tclENV play nicely with abstract?

Unclear even to me, but worth testing.

> What is the "VM-acting function" you refer to?

The Csymtable. I did not realize that this function was calling `add_leaf` which makes it even worse. No tactic should mess with the libobject stack, the only one I thought was doing that was precisely abstract.

> Are you suggesting we instead do something like

If this works with abstract, then yes. Otherwise, just use the `Global.set_strategy` function, or even better maybe both this and the `tclENV` solution to make the code future proof.

Reply:
@ppedrot I tried this:
```ocaml
{

let set_strategy lvl ql =
  let glob_key r =
    match r with
    | GlobRef.ConstRef sp -> ConstKey sp
    | GlobRef.VarRef id -> VarKey id
    | _ -> user_err Pp.(str
        "cannot set an inductive type or a constructor as transparent") in
  Proofview.tclENV >>= fun env ->
  let env = List.fold_left (fun env q ->
    Environ.set_oracle env
      (Conv_oracle.set_strategy (Environ.oracle env) (glob_key q) lvl)) env ql in
  Proofview.Unsafe.tclSETENV env

}

TACTIC EXTEND strategy
| [ "strategy" strategy_level_or_var(v) "[" ne_smart_global_list(q) "]"  ] ->
  { set_strategy v q }
END
```
and it does not work so well.  It works for `strategy transparent [id]; unfold id`, but fails for `strategy transparent [id]. unfold id.`  How do I set the tactic environment in a way that persists across periods?

Reply:
@ppedrot I've updated the code to call `Global.set_strategy` more directly, which does seem to work fine.

Reply:
@JasonGross unfortunately in the current design you cannot update the global env from tactic code, this will create problem for example with workers. If you need to set the strategy in tactics that bit of state will have to be moved to the proof state.

Reply:
@ejgallego Sure, I noticed that if I do `Opaque id. par: strategy transparent [id]; shelve. Unshelve. unfold id.`, that fails. (I don't think there's any other way to get two bits of the same proof to be spread across workers, right?)  You're saying it's bad to commit this tactic when things are in such a state, because it creates technical debt, even if the tactic is already useful even if it doesn't work well with `par:`?  (I could insert a notice in the manual that this tactic doesn't play well with `par:`, for the time being.)

Indeed, setting the strategy in tactics is quite important (and eventually I want a way to make it persist long enough that it is still in effect during the running of `Qed`), so I think this should be moved to the proof state.

Reply:
I'd say let's move it to the proof state now [if the rest of the devs agree], IMVHO we have enough "corner cases" in Coq as to create more inconsistencies.

Specially when we the KNOWN_ISSUES file was rejected so we don't even document them.

Reply:
> I'd say let's move it to the proof state now

Sounds good.  Is this a PR you're willing to make, or a change I should try to make (and if me, should it be in this PR, or a separate one)?

Reply:
We can make it in a different PR, let's hear from @ppedrot tho first.

Reply:
I am not sure I understand the plan. The environment is already part of the proof state, and it contains the oracle state as well. What do you want to add?

Reply:
Not really, the environ that is part of the proof state is not the `Global` env, so effectively you have two environs; as pointed out here unless you call `Global.set_oracle` the changes won't be preserved for example across `run_tactic`.